project_W/model.py

```diff
@@ -215,18 +237,130 @@ def _send_email(
+        file=InputFile(
+            file_name=file_name,
+            audio_data=audio
+        ),
+        model=model,
+        language=language,
+    )
+
+    db.session.add(job)
+    db.session.commit()
+
+    logger.info(f"User {user.email} submitted job with file {file_name}")
+    logger.info(f" -> Assigned job id {job.id}")
+
+    return job
+
+
+def get_job_by_id(id: int) -> Optional[Job]:
+    return db.session.query(Job).where(Job.id == id).one_or_none()
+
+
+def list_job_ids_for_user(user: User) -> List[int]:
+    """Returns a list of all the job IDs associated with this user"""
+    return [job.id for job in user.jobs]
+
+
+def list_job_ids_for_all_users() -> Dict[int, List[int]]:
+    """For each user, returns a list of all job IDs associated with that user"""
+    return {user.id: list_job_ids_for_user(user) for user in db.session.query(User)}
+
+
+def runner_token_hash(token: str) -> str:
+    return base64.urlsafe_b64encode(
+        hashlib.sha256(token.encode("ascii")).digest()
+    ).rstrip(b"=").decode("ascii")
+
+
+@dataclass
+class Runner(db.Model):
+    __tablename__ = "runners"
+    id = db.Column(db.Integer, primary_key=True)
+    # We only store the hash of the token, otherwise a db leak would make
+    # it possible to impersonate any runner. We don't need to use a salted hash
+    # because the token is created by the server and already has sufficient entropy.
+    # The hash itself is stored using base64.
+    token_hash = db.Column(db.Text, nullable=False, unique=True, index=True)
+    # TODO: Add some form of runner tag system
+
+
+def get_runner_by_token(token: str) -> Optional[Runner]:
+    token_hash = runner_token_hash(token)
+    return db.session.query(Runner).where(Runner.token_hash == token_hash).one_or_none()
+
+
+def create_runner() -> Tuple[str, Runner]:
+    token = secrets.token_urlsafe()
+    token_hash = runner_token_hash(token)
+
+    # Sanity check to ensure that the token and its hash are unique.
+    while db.session.query(Runner).where(Runner.token_hash == token_hash).one_or_none():
+        token = secrets.token_urlsafe()
+        token_hash = runner_token_hash(token)
+
+    runner = Runner(token_hash=token_hash)
+    db.session.add(runner)
+    db.session.commit()
+    return token, runner
+
+
+# some additional wraps for api routes
+# if user should be activated to use this route
 def activatedRequired(f):
     @wraps(f)
     def decoratedFunction(*args, **kwargs):
         user: User = current_user
         if not user.activated:
             return jsonify(msg="Your account is not activated. Please activate your user account to use this feature.", errorType="permission"), 403
-        else: return f(*args, **kwargs)
+        else:
+            return f(*args, **kwargs)
     return decoratedFunction
 
-#if user should need to confirm their identity by entering their password
+# if user should need to confirm their identity by entering their password
+
+
 def confirmIdentity(f):
```

Comment:
these are also automatically inserted by autopep8. I'm not sure why it's doing that, (and it also doesn't always do it, e.g. on `activatedRequired`), but I'm guessing it's because autopep8 doesn't consider these to be doc comments. I guess I'll change them to be doc strings instead.